---
title: 'CoRe: Identification of core-fitness genes in genome-wide CRISPR-Cas9 screens'
author: "Emre Karakoc, Clare Pacini, Alessandro Vinceti and Francesco Iorio"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_depth: '3'
    df_print: paged
  rmarkdown::html_vignette:
    toc: yes
    toc_depth: 3
vignette: |
  %\VignetteIndexEntry{Vignette Title} %\VignetteEngine{knitr::rmarkdown} %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
```

<br>

![](../web/coRe_logo.jpg)

<br> <br> <br><br>
[CoRe R package interactive vignette](https://rpubs.com/AleVin1995/CoRe)
<br> <br> <br><br>

## Introduction

The CoRe package implements two methods for identification of core fitness genes (at two level of stringency) from joint analyses of multiple genome-wide CRISPR-Cas9 screens: 

1) The percentile ranking method builds on the starting basic intuition that if a gene is involved in essential processes in every cell, it should be among the top essential genes in all the analysed screens, even those performed on the cell lines whose viability is lowly impacted by the knock-out of that gene, i.e. the least dependent cell lines. The CoRe package implements three variants of this method, which verify if the starting assumption is satisfied for (considering each gene in turn) in different ways. The first variant accounts for the distribution of the rank position of the genes in their least dependent cell lines, obtained by sorting all genes based on their essentiality in that cell line. The second considers the distribution of the average gene rank positions in all the cell lines falling at least in 90-th percentile cell lines of least dependent ones.
The third version, fits a linear model on the curve obtained when sorting all cell lines based on how much they are dependent on a given gene then considering the essentiality rank position of that gene across all cell lines (in the obtained order).
The density of the considered distributions is then estimated using a Gaussian kernel with width 0.1 and
the point of minimum density identified. Genes falling below the point of minimum density are classified as common
essential. Alternatively, two normal distributions are fitted and the best separation threshold is determined. In this case, genes whose rank position, or average rank position, or slope is predicted to be generated by the first distribution are called common essential genes, i.e. core fitness genes at a lower stringency [1]. 

2) The Adaptive Daisy Model (ADaM) is a semi-supervised algorithm for computing a fuzzy-intersection of non-fuzzy sets by adaptively determining the minimal number of sets to which an element should belong in order to be a member of the fuzzy-intersection (the membership threshold). This threshold maximises the deviance from expectation of the cardinality of the resulting fuzzy-intersection, as well as the coverage of predefined elements.
This method can be used to identify the minimal number of cell lines from a given tissue in which the inactivation of a gene (via CRISPR-Cas9 targeting) should exert a reduction of viability (or fitness effect) in order for that gene to be considered a core-fitness essential gene for the tissue under consideration. Iterating this method and considering cancer-type specific core fitness genes as sets to be fuzzy-intersected, then this method can be used to estimate a set of pan-cancer core fitness genes.
This method is used to discriminate between core-fitness and context-specific essential genes in a study describing a large scale genome-wide CRISPR-Cas9 pooled drop-out screening [1] (a detailed description of the algorithm is included in the Supplemental Information of [2]). ADaM was inspired by the Daisy Model method introduced in [3]

License: GNU GPL v3

Contributors: Emre Karakoc, Clare Pacini, Alessandro Vinceti and Francesco Iorio

**References**

[1] J. M. Dempster et al., Agreement between two large pan-cancer CRISPR-Cas9 gene dependency data sets., Nat. Commun., vol. 10, no. 1, p. 5817, 2019, doi: 10.1038/s41467-019-13805-y.

[2] Behan FM & Iorio F & Picco G et al., Prioritisation of cancer therapeutic targets using CRISPR-Cas9 screens. Nature, In press.

[3] Hart T et al., High-Resolution CRISPR Screens Reveal Fitness Genes and Genotype-Specific Cancer Liabilities. Cell. 2015;163:1515â€“26.

## Running Modalities

CoRe is an R package available at: https://github.com/DepMap-Analytics/CoRe.

This page contains instruction to quickly try the package. User manual and package documentation are available at
https://github.com/DepMap-Analytics/CoRe/blob/master/Reference_Manual.pdf.

## R package: quick start

### Package installation

The R package is available on github at https://github.com/DepMap-Analytics/CoRe. We recommend to use it within Rstudio (https://www.rstudio.com/). To install it the following commands should be executed:

```{r, fig.show='hold', eval=FALSE, results='hide'}
library(devtools)
install_github("DepMap-Analytics/CoRe")
library(CoRe)
```

This will install the following additional libraries:
```{r, fig.show='hold', eval=FALSE, results='hide'}
mixtools, magrittr, RCurl, readr, pheatmap, stringr
```

all publicly available on CRAN. <br><br>
The package comes with built-in data objects containing predefined sets of BAGEL essential and non-essential genes (from _Hart et al_, Cell 2015), curated BAGEL genes (as defined in _Behan et al_, Nature 2019) other sets of a priori known essential genes (from http://www.gsea-msigdb.org/gsea/msigdb/index.jsp).

### Execute ADaM pipeline in CoRe package step by step

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
library(CoRe)

## Downloading binary dependency matrix
## for > 300 cancer cell lines from Project Score
BinDepMat<-CoRe.download_BinaryDepMatrix()

## Extracting dependency submatrix for
## Non-Small Cell Lung Carcinoma cell lines only
LungDepMap<-CoRe.extract_tissueType_SubMatrix(BinDepMat,
                                              tissue_type="Non-Small Cell Lung Carcinoma")

# Generate the profiles of number of fitness genes across number of cell lines from observed data and
# corresponding cumulative sums.
pprofile<-CoRe.panessprofile(depMat=LungDepMap)

# Generate a set of random profiles of number of genes depleted for a number of cell lines and corresponding
# cumulative sums by perturbing observed data.
nullmodel<-CoRe.generateNullModel(depMat=LungDepMap,
                                  ntrials = 1000)

# Calculate log10 odd ratios of observed/expected profiles of cumulative number of fitness genes in fixed number of cell lines
EO<-CoRe.empiricalOdds(observedCumSum = pprofile$CUMsums,
                       simulatedCumSum =nullmodel$nullCumSUM)

# Load a reference set of essential genes
# (For others set of essential genes are available: type "??CoRe::datasets")
data(curated_BAGEL_essential)

# Calculate True positive rates for fitness genes in at least n cell lines in the observed dependency matrix,
# with positive cases from a reference set of essential genes
TPR<-CoRe.truePositiveRate(LungDepMap,
                           curated_BAGEL_essential)

# Calculate minimum number of cell lines a gene needs to be a fitness gene in order to be considered
# as a core-fitness gene
crossoverpoint<-CoRe.tradeoffEO_TPR(EO,
                                    TPR$TPR,
                                    test_set_name = 'curated BAGEL essential')

# coreFitnessGenes is the list of genes predicted as core-fitness by AdAM.
coreFitnessGenes<-CoRe.coreFitnessGenes(LungDepMap,
                                        crossoverpoint)
```

### Execute previous pipeline using ad-hoc CoRe function for specific tissue/cancer types

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
# Reperform the same pipeline using CoRe.CS_ADaM function
clannotation<-CoRe.download_AnnotationModel()

SNCLC_CFgenes<-CoRe.CS_AdAM(BinDepMat,tissue_ctype = 'Non-Small Cell Lung Carcinoma',
                             clannotation = clannotation,
                             TruePositives = curated_BAGEL_essential)
```

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
BRCA_CFgenes<-CoRe.CS_AdAM(BinDepMat,tissue_ctype = 'Breast',
                            clannotation = clannotation,
                            TruePositives = curated_BAGEL_essential)
```

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
CRC_CFgenes<-CoRe.CS_AdAM(BinDepMat,tissue_ctype = 'Large Intestine',
                           clannotation = clannotation,
                           TruePositives = curated_BAGEL_essential)
```

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
################################################################################
#                                                                              #
#  WARNING: For the ADaM method to work properly we recommend to execute the   #
#  pipeline on tissue/cancer types with at least 8 cell lines.                 #
#                                                                              #
################################################################################

# Cell Model Passport cancer tissue with at least 8 cell lines
names(which(table(clannotation$tissue) >= 8))
```

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
# Cell Model Passport cancer types with at least 8 cell lines
names(which(table(clannotation$cancer_type) >= 8))
```

### Execute AdAM method to compute Pancancer core-fitness genes

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
# Identifying pan-cancer core-fitness genes with the AdAM model, as
# described in Behan et al 2019, i.e. performing analyses at individual
# tissues/cancer-type level then collapsing results at pan-cancer level

# Tissue/cancer types
tissues_ctypes<-c("Haematopoietic and Lymphoid",
                  "Ovary",
                  "Peripheral Nervous System",
                  "Central Nervous System",
                  "Pancreas",
                  "Head and Neck",
                  "Bone",
                  "Lung",
                  "Large Intestine",
                  "Esophagus",
                  "Endometrium",
                  "Stomach",
                  "Breast")

clannotation<-
  CoRe.download_AnnotationModel('https://cog.sanger.ac.uk/cmp/download/model_list_latest.csv.gz')

PanCancer_CFgenes<-
  CoRe.PanCancer_AdAM(pancan_depMat = BinDepMat,
                      tissues_ctypes = tissues_ctypes,
                      clannotation = clannotation,
                      TruePositives = curated_BAGEL_essential,
                      display = FALSE)

head(PanCancer_CFgenes)

gene<-PanCancer_CFgenes[1]

# Downloading quantitative dependency matrix
# from Project Score
depMat<-CoRe.download_DepMatrix(scaled = TRUE, 
                                ess = curated_BAGEL_essential, 
                                noness = curated_BAGEL_nonEssential)

# Visualization of CFness of a gene and comparison 
# to positive and negative control genes
CoRe.VisCFness(depMat,
               gene,
               percentile=0.9,
               posControl='RPL12',
               negControl='MAP2K1')
```

### Execute 90-th Percentile method in CoRe package

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
# Load reference sets of essential and non-essential genes
# (Others reference sets of genes are available: type "??CoRe::datasets")
data(curated_BAGEL_essential)
data(curated_BAGEL_nonEssential)

# Prior known sets of independent essential genes
data(EssGenes.DNA_REPLICATION_cons)
data(EssGenes.HISTONES)
data(EssGenes.KEGG_rna_polymerase)
data(EssGenes.PROTEASOME_cons)
data(EssGenes.SPLICEOSOME_cons)
data(EssGenes.ribosomalProteins)

signatures<-list(DNA_REPLICATION=EssGenes.DNA_REPLICATION_cons,
                 HISTONES=EssGenes.HISTONES,
                 RNA_POLYMERASE=EssGenes.KEGG_rna_polymerase,
                 PROTEASOME=EssGenes.PROTEASOME_cons,
                 SPLICEOSOME=EssGenes.SPLICEOSOME_cons,
                 RIBOSOMAL_PROTS=EssGenes.ribosomalProteins)

# Download CCLE gene expression data from DepMap portal
FPs<-CoRe.AssembleFPs()

################################################################################
#                                                                              #
#  IMPORTANT: To execute the 90-th percentile method on a specific tissue or   #
#  cancer type, use CoRe.extract_tissueType_SubMatrix function to extract the  #
#  submatrix.                                                                  #
#                                                                              #
################################################################################

# 90-th Percentile Method using different methods and thresholds
CFgenes<-CoRe.PercentileCF(depMat,
                           method = 'fixed',
                           thresholding='localMin')
```

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
CFgenesAVG<-CoRe.PercentileCF(depMat,
                              method = 'average',
                              thresholding='localMin')
```

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
CFgenesSLOPE<-CoRe.PercentileCF(depMat,
                                method = 'slope',
                                thresholding='localMin')
```

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
CFgenes_BFs<-CoRe.PercentileCF(depMat,
                               method = 'fixed',
                               thresholding='BFs')
```

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
CFgenesAVG_BFs<-CoRe.PercentileCF(depMat,
                                  method = 'average',
                                  thresholding='BFs')
```

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
CFgenesSLOPE_BFs<-CoRe.PercentileCF(depMat,
                                    method = 'slope',
                                    thresholding='BFs')
```

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
# Benchmarking the identified Pancancer core-fitness genes against
# prior known essential genes
AdAMperf<-CoRe.CF_Benchmark(CFgenes$cfgenes,
                            background = rownames(depMat),
                            priorKnownSignatures = signatures,
                            falsePositives=FPs)
```

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
AdAMperfAVG<-CoRe.CF_Benchmark(CFgenesAVG$cfgenes,
                                background = rownames(depMat),
                                priorKnownSignatures = signatures,
                                falsePositives=FPs)
```

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
AdAMperfSLOPE<-CoRe.CF_Benchmark(CFgenesSLOPE$cfgenes,
                                  background = rownames(depMat),
                                  priorKnownSignatures = signatures,
                                  falsePositives=FPs)
```

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
AdAMperf_BFs<-CoRe.CF_Benchmark(CFgenes_BFs$cfgenes,
                                background = rownames(depMat),
                                priorKnownSignatures = signatures,
                                falsePositives=FPs)
```

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
AdAMperfAVG_BFs<-CoRe.CF_Benchmark(CFgenesAVG_BFs$cfgenes,
                                    background = rownames(depMat),
                                    priorKnownSignatures = signatures,
                                    falsePositives=FPs)
```

```{r, fig.show='hold', eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
AdAMperfSLOPE_BFs<-CoRe.CF_Benchmark(CFgenesSLOPE_BFs$cfgenes,
                                      background = rownames(depMat),
                                      priorKnownSignatures = signatures,
                                      falsePositives=FPs)
```
