{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CoRe package benchmarking\n",
    "\n",
    "We provide a systematic assessment of the core fitness genes identified by the methods implemented in the CoRe package, namely the 90-th percentile ranking method and the Adaptive Daisy Model (ADaM).\n",
    "As testing, we use the CERES integrated dependency dataset (release 1 Sanger, version 20Q2 Broad). It has been shown in [1] that CERES provides a dataset optimized for the identification of a priori known essential and non-essential genes. The CERES dataset is scaled using curated BAGEL essential and non-essential genes [2], so that the median of the former is equal to -1 and the median of the latter is equal to 0 for every cell line.\n",
    "Since ADaM works with binary data, we binarized the CERES matrix considering all the instances having a fold-change score less than -0.5 as essential, otherwise they are deemed non-essential.\n",
    "\n",
    "__References__ <br>\n",
    "[1] C. Pacini et al., “Integrated cross-study datasets of genetic dependencies in cancer”, Nat. Commun., vol. 12, no. 1, p. 1661, 2021, doi: 10.1038/s41467-021-21898-7. <br>\n",
    "[2] F. M. Behan et al., “Prioritization of cancer therapeutic targets using CRISPR–Cas9 screens”, Nature, 2019, doi: 10.1038/s41586-019-1103-9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "\n",
    "library(tidyverse)\n",
    "library(pheatmap)\n",
    "library(CoRe)\n",
    "library(magrittr)\n",
    "\n",
    "## curated sets of BAGEL genes\n",
    "data('curated_BAGEL_essential')\n",
    "data('curated_BAGEL_nonEssential')\n",
    "\n",
    "## CERES scaling\n",
    "scale_to_essentials <- function(ge_fit,ess_genes,noness_genes){\n",
    "  essential_indices <- which(row.names(ge_fit) %in% ess_genes)\n",
    "  nonessential_indices <- which(row.names(ge_fit) %in% noness_genes)\n",
    "  scaled_ge_fit <- ge_fit %>%\n",
    "    apply(2, function(x){\n",
    "      (x - median(x[nonessential_indices], na.rm=T)) %>%\n",
    "        divide_by(median(x[nonessential_indices], na.rm=T) - median(x[essential_indices], na.rm=T))\n",
    "    })\n",
    "  return(scaled_ge_fit)\n",
    "}\n",
    "\n",
    "## download CERES integrated dataset\n",
    "url <- 'https://www.depmap.org/broad-sanger/integrated_Sanger_Broad_essentiality_matrices_20201201.zip'\n",
    "temp <- tempfile()\n",
    "download.file(url, temp, mode=\"wb\") \n",
    "unzip(temp, exdir = 'integrated_dataset')\n",
    "unlink(temp)\n",
    "\n",
    "depFC <- read.table('integrated_dataset/integrated_Sanger_Broad_essentiality_matrices_20201201/CERES_FC.txt', \n",
    "                    row.names = 1, sep = '\\t', header = TRUE, stringsAsFactors = FALSE, check.names = FALSE)\n",
    "system('rm -r integrated_dataset')\n",
    "\n",
    "## download cell line annotation from the Cell Model Passports\n",
    "clannotation <- read_csv('data/model_list_20210326.csv')\n",
    "\n",
    "## convert model IDs to model names\n",
    "depFC <- depFC[,-which(is.na(colSums(depFC)))]\n",
    "cells <- colnames(depFC)\n",
    "\n",
    "tot_id <- clannotation %>% filter(model_id %in% cells | BROAD_ID %in% cells) %>% \n",
    "  select(model_name,model_id,BROAD_ID) %>% \n",
    "  pivot_longer(cols = c(\"model_id\",\"BROAD_ID\"), names_to = \"institute\", values_to = \"model_id\") %>% \n",
    "  select(model_name,model_id) %>% filter(model_id %in% cells) %>% \n",
    "  arrange(factor(model_id, levels = cells))\n",
    "\n",
    "depFC <- depFC[,which(cells %in% tot_id$model_id)]\n",
    "colnames(depFC) <- tot_id$model_name\n",
    "scaled_depFC <- scale_to_essentials(depFC,curated_BAGEL_essential,curated_BAGEL_nonEssential)\n",
    "\n",
    "bdep <- apply(scaled_depFC, 2, function(x){\n",
    "  x[which(x >= -0.5)] <- 0\n",
    "  x[which(x < -0.5)] <- 1\n",
    "  x\n",
    "}) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computation of pan-cancer core-fitness genes\n",
    "\n",
    "In this section, we compute the pan-cancer core-fitness genes using ADaM with cancer tissues constituted by at least 15 cell lines and the four variants of the 90-th percentile ranking method.\n",
    "For a detailed description of the methods, please visit the github page: https://github.com/DepMap-Analytics/CoRe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## determining tissues with at least 15 cell lines in the dependency matrix\n",
    "tissues_ctypes<-\n",
    "  names(which(summary(as.factor(clannotation$tissue[match(colnames(bdep),clannotation$model_name)]))>=15))\n",
    "\n",
    "print('ADaM will be executed at the tissue type level on the following tissue lineages:')\n",
    "print(tissues_ctypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ADaM\n",
    "ADaM <- CoRe.PanCancer_ADaM(bdep,\n",
    "                            tissues_ctypes,\n",
    "                            clannotation = clannotation,\n",
    "                            display=FALSE,\n",
    "                            ntrials=1000,\n",
    "                            verbose=FALSE,\n",
    "                            TruePositives = curated_BAGEL_essential)\n",
    "\n",
    "## 90-th percentile with fixed strategy\n",
    "Perc_fixed <- CoRe.FiPer(scaled_depFC,\n",
    "                         display=FALSE,\n",
    "                         method = 'fixed')$cfgenes\n",
    "\n",
    "## 90-th percentile with avg strategy\n",
    "Perc_avg <- CoRe.FiPer(scaled_depFC,\n",
    "                       display=FALSE,\n",
    "                       method = 'average')$cfgenes\n",
    "\n",
    "## 90-th percentile with slope strategy\n",
    "Perc_slope <- CoRe.FiPer(scaled_depFC,\n",
    "                         display=FALSE,\n",
    "                         method = 'slope')$cfgenes\n",
    "\n",
    "## 90-th percentile with AUC strategy\n",
    "Perc_AUC <- CoRe.FiPer(scaled_depFC,\n",
    "                       display=FALSE,\n",
    "                       method = 'AUC')$cfgenes\n",
    "\n",
    "## common core-fitness genes\n",
    "Perc_Consensus <- Reduce(intersect,list(Perc_fixed,Perc_avg,Perc_slope,Perc_AUC))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary statistics\n",
    "\n",
    "To extend our comparison, we also decided to include the core-fitness (CF) genes predicted by the ADaM method in [1], by CEN-tools in [2] and by CEN-tools using the curated BAGEL genes as training sets on the CERES dataset we just computed. Also the Hart essential genes in [3] and the new version presented in [4] are taken into account. <br>\n",
    "First, we show the number of core-fitness genes for each set under consideration and then we compute the similarity among them.\n",
    "\n",
    "__References__ <br>\n",
    "[1] F. M. Behan et al., “Prioritization of cancer therapeutic targets using CRISPR–Cas9 screens”, Nature, 2019, doi: 10.1038/s41586-019-1103-9. <br>\n",
    "[2] S. Sharma, C. Dincer, P. Weidemüller, G. J. Wright, and E. Petsalaki, “<scp>CEN</scp>‐tools: an integrative platform to identify the contexts of essential genes”, Mol. Syst. Biol., vol. 16, no. 10, Oct. 2020, doi: 10.15252/msb.20209698. <br>\n",
    "[3] T. Hart et al., “High-Resolution CRISPR Screens Reveal Fitness Genes and Genotype-Specific Cancer Liabilities”, Cell, vol. 163, no. 6, pp. 1515–26, Dec. 2015, doi: 10.1016/j.cell.2015.11.015. <br>\n",
    "[4]\tT. Hart et al., “Evaluation and Design of Genome-Wide CRISPR/SpCas9 Knockout Screens”, G3, Genes|Genomes|Genetics, vol. 7, no. 8, pp. 2719–2727, Aug. 2017, doi: 10.1534/g3.117.041277."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Essentials genes obtained with CENtools using BAGEL curated genes as reference sets (default nº bin = 20)\n",
    "load('data/CENtools_Essentials.RData')\n",
    "\n",
    "## CENtools\n",
    "CENtools_SV1_B20Q2_CERES <- CENtools_ess\n",
    "\n",
    "## CENtools Sharma et al, 2020\n",
    "CENtools_2020<-read.table(file = 'data/CenTools_essential_Sharma_et_al.txt',sep='\\t',stringsAsFactors = FALSE)$V1\n",
    "\n",
    "## ADaM essential Behan et al., 2019\n",
    "load('data/ADaM_CFs_Behan_et_Al_2019.RData')\n",
    "ADaM_2019<-PanCancerCoreFitnessGenes\n",
    "\n",
    "## Hart et al., 2015\n",
    "data(BAGEL_essential)\n",
    "Hart_Essential_2015<-BAGEL_essential\n",
    "\n",
    "## Hart et al., 2017\n",
    "load('data/BAGEL_v2_Essentials.RData')\n",
    "Hart_Essential_2017<-BAGEL_essential\n",
    "\n",
    "##### Barplot comparing n. of predicted CF genes\n",
    "CFs_sets<-list(Hart_2015 = Hart_Essential_2015,\n",
    "               Hart_2017 = Hart_Essential_2017,\n",
    "               Behan_2019 = ADaM_2019,\n",
    "               ADaM = ADaM,\n",
    "               F_Perc_avg = Perc_avg,\n",
    "               F_Perc_Cons_lax = Perc_Consensus,\n",
    "               F_Perc_slope = Perc_slope,\n",
    "               F_Perc_AUC = Perc_AUC,\n",
    "               F_Perc_fixed = Perc_fixed)\n",
    "\n",
    "col=c(\"#03B2C8\",\"#034DD9\",\"#E6AB02\",\"#1B9E77\",\"#337100\",\"#FC8D62\",'black',\n",
    "      \"#F4CAE4\",'purple',\"#CAB2D6\",\"#BEBADA\",\"#777892\")\n",
    "names(col)<-c(\"Hart_2015\",\"Hart_2017\",\"Behan_2019\",\"Sharma_2020\",\"CENtools\",\"ADaM\",\n",
    "              \"F_Perc_Cons\",\"F_Perc_avg\",\"F_Perc_Cons_lax\",\"F_Perc_slope\",\"F_Perc_AUC\",\"F_Perc_fixed\")\n",
    "\n",
    "par(mar=c(8,6,4,4))\n",
    "barplot(unlist(lapply(CFs_sets,length)),ylab='n. genes',las=2,border = FALSE,col=col[names(CFs_sets)],\n",
    "        cex.lab = 1.4, cex.main = 2, ylim = c(0,2500))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### CF gene set similarity\n",
    "allEss<-unique(unlist(CFs_sets))\n",
    "membMat<-do.call(rbind,lapply(CFs_sets,function(x){is.element(allEss,x)}))\n",
    "rownames(membMat)<-names(CFs_sets)\n",
    "colnames(membMat)<-allEss\n",
    "membMat<-membMat+0\n",
    "\n",
    "pheatmap(membMat,show_colnames = FALSE,clustering_distance_rows = 'binary',cluster_cols = FALSE,border_color = NA,\n",
    "         main = 'Similarity across sets')\n",
    "\n",
    "dmat<-dist(membMat,method = 'binary')\n",
    "\n",
    "pheatmap(1-as.matrix(dmat), main = 'JS for Pan-cancer core fitness genes across methods', \n",
    "         legend = FALSE,display_numbers = round(dmat,digits = 2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Baseline predictor\n",
    "\n",
    "A baseline predictor is defined by considering CF genes those essential in at least n cell lines for all n. We compute the recall of this predictor using independent sets of a priori known essential genes (i.e. those involved in housekeeping cellular processes like translation or replication) [1-2]. <br>\n",
    "In addition, gene dependencies associated with a biomarker are very specific to the molecular context, therefore, these genes classified as CF are deemed false positives (FP). In the FP set are included also the non-expressed genes. We compute the recall of these FP genes for the baseline predictor.\n",
    "\n",
    "__References__ <br>\n",
    "[1]\tF. Iorio et al., “Unsupervised correction of gene-independent cell responses to CRISPR-Cas9 targeting,” BMC Genomics, 2018, doi: 10.1186/s12864-018-4989-y. <br>\n",
    "[2]\tC. Pacini et al., “Integrated cross-study datasets of genetic dependencies in cancer,” Nat. Commun., vol. 12, no. 1, p. 1661, 2021, doi: 10.1038/s41467-021-21898-7."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### Ability to identify novel hits\n",
    "CFs_sets<-list(Behan_2019 = ADaM_2019,\n",
    "               Sharma_2020 = CENtools_2020,\n",
    "               CENtools = CENtools_SV1_B20Q2_CERES,\n",
    "               ADaM = ADaM,\n",
    "               F_Perc_avg = Perc_avg,\n",
    "               F_Perc_Cons_lax = Perc_Consensus,\n",
    "               F_Perc_slope = Perc_slope,\n",
    "               F_Perc_AUC = Perc_AUC,\n",
    "               F_Perc_fixed = Perc_fixed)\n",
    "\n",
    "## Assembling a set of prior known essential genes that are not included in the\n",
    "## training sets used by CENtools and/or ADaM\n",
    "data(EssGenes.DNA_REPLICATION_cons)\n",
    "data(EssGenes.KEGG_rna_polymerase)\n",
    "data(EssGenes.PROTEASOME_cons)\n",
    "data(EssGenes.SPLICEOSOME_cons)\n",
    "data(EssGenes.ribosomalProteins)\n",
    "data(EssGenes.HISTONES)\n",
    "\n",
    "signatures<-list(DNA_REPLICATION=EssGenes.DNA_REPLICATION_cons,\n",
    "                 RNA_POLYMERASE=EssGenes.KEGG_rna_polymerase,\n",
    "                 PROTEASOME=EssGenes.PROTEASOME_cons,\n",
    "                 SPLICEOSOME=EssGenes.SPLICEOSOME_cons,\n",
    "                 RIBOSOMAL_PROTS=EssGenes.ribosomalProteins,\n",
    "                 HISTONES=EssGenes.HISTONES)\n",
    "\n",
    "signatures<-unique(unlist(signatures))\n",
    "\n",
    "geneSet <- rownames(scaled_depFC)\n",
    "\n",
    "## Adding additional signatures in Pacini et al, 2020\n",
    "load(\"data/Kegg.DNArep.Rdata\")\n",
    "load(\"data/Kegg.Ribosome.Rdata\")\n",
    "load(\"data/Kegg.Proteasome.Rdata\")\n",
    "load(\"data/Kegg.Spliceosome.Rdata\")\n",
    "load(\"data/Kegg.RNApoly.Rdata\")\n",
    "load(\"data/Histones.Rdata\")\n",
    "\n",
    "allRefEss<-unique(c(Kegg.DNArep,Kegg.Proteasome,Kegg.Ribosome,Kegg.RNApoly,Kegg.Spliceosome,Histones))\n",
    "\n",
    "signatures<-union(signatures,allRefEss)\n",
    "signatures <- intersect(signatures,geneSet)\n",
    "\n",
    "## training sets used by ADaM/CENtools while executed on the new joint dataset\n",
    "## and by ADaM on the old dataset\n",
    "BAG_genes <- unique(c(curated_BAGEL_essential,\n",
    "                      curated_BAGEL_nonEssential,\n",
    "                      BAGEL_essential,\n",
    "                      BAGEL_nonEssential))\n",
    "\n",
    "## removing training sets used by ADaM/CENtools while executed on the new joint dataset\n",
    "## and by ADaM on the old dataset\n",
    "signatures <- setdiff(signatures,BAG_genes)\n",
    "geneSet_noBAG <- setdiff(geneSet,BAG_genes)\n",
    "\n",
    "## baseline predictor\n",
    "baselineCFs <- lapply(1:ncol(bdep),function(x){\n",
    "  names(which(rowSums(bdep)>=x))\n",
    "})\n",
    "\n",
    "baselineSizes<-unlist(lapply(baselineCFs,length))\n",
    "\n",
    "baselineRecalls<-100*unlist(lapply(baselineCFs,function(x){\n",
    "  length(intersect(x,signatures))/length(signatures)\n",
    "}))\n",
    "\n",
    "load('data/CMP_RNAseq.Rdata')\n",
    "\n",
    "## genes not expressed\n",
    "lowlyExp<-names(which(rowSums(CMP_RNAseq<0.01,na.rm=TRUE)>=ncol(CMP_RNAseq)))\n",
    "\n",
    "## dependency associated with biomarkers\n",
    "FPs <- sort(unique(unlist(read.table('data/dependency_with_biomarkers.txt',stringsAsFactors = FALSE))))\n",
    "FPs <- union(FPs,lowlyExp)\n",
    "FPs <- intersect(FPs,geneSet_noBAG)\n",
    "\n",
    "baselineNEratio<-100*unlist(lapply(baselineCFs,function(x){\n",
    "  length(intersect(x,FPs))/length(FPs)\n",
    "}))\n",
    "\n",
    "plot(baselineSizes,xlab='Baseline predictor of CF genes (essentials in >= n cell lines)', \n",
    "     ylab='n. of predicted CF genes',pch=16,log='y',main = 'BAGEL genes removed')\n",
    "\n",
    "plot(baselineRecalls,xlab='Baseline predictor of CF genes (essentials in >= n cell lines)',\n",
    "     ylab='% Recall of prior known essential genes not used for training',pch=16)\n",
    "\n",
    "plot(baselineNEratio,xlab='Baseline predictor of CF genes (essentials in >= n cell lines)',\n",
    "     ylab='% Recall of FP genes (associated with a biomarker or lowly expressed)',pch=16)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison between method considering novel hits\n",
    "\n",
    "The metrics derived from the baseline predictors are then used to assess the CF sets predicted by each method. We consider the CF sets without the BAGEL genes used for the training of CEN-tools, namely BAGEL essential version 2 and non-essential used for Sharma_2020 set, and curated BAGEL essential and non-essential used for CENtools set. <br>\n",
    "In the first plot, we normalised the recall of the signatures by the maximum recall achieved by the baseline predictor. The 90-th percentile variants have a higher ratio since they are less stringent. <br>\n",
    "In the second plot, we show the recall ratio of the FP genes. ADaM and CEN-tools tend to show a higher stringency than the 90-th percentile variants, though the 'average' variant ratio is close to the ADaM and CEN-tools sets. <br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fCFs_sets <- lapply(CFs_sets,function(x){\n",
    "  setdiff(x,BAG_genes)\n",
    "})\n",
    "\n",
    "observed_recalls<-100*unlist(lapply(fCFs_sets,\n",
    "                                    function(x){length(intersect(x,signatures))/length(signatures)}))\n",
    "\n",
    "observed_FPs<-100*unlist(lapply(fCFs_sets,\n",
    "                                function(x){length(intersect(x,FPs))/length(FPs)}))\n",
    "\n",
    "par(mar=c(8,6,4,4))\n",
    "barplot(observed_recalls/max(baselineRecalls),col=col[names(observed_recalls)],\n",
    "        las=2,ylab='Recall / (max Recall)',border=NA,main = 'Prior known essential not included in training')\n",
    "\n",
    "barplot(observed_FPs/max(baselineNEratio),col=col[names(observed_FPs)],\n",
    "        las=2,ylab='Recall / (max Recall)', border=NA, main = 'False positive CFs')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "However, in order to have a fair comparison between the methods, we need to compare both the recall of prior known essential genes (from here on referred as true positive, TP) and the recall of FP genes to the baseline predictor. To do so we draw the relationship between TP and FP genes for the baseline predictor and show where each method stand. This approach shows that the 90-th percentile variants are more stringent than CEN-tools and ADaM when the size of predicted CF genes is taken into account."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot(spline(baselineRecalls,baselineNEratio),pch=16,\n",
    "     xlab='% Recall of prior known essentials',\n",
    "     ylab='% Recall of selective essentials (associated with a biomarker)',type='l',lwd=4,\n",
    "     xlim=c(10,50),ylim=c(0,6))\n",
    "\n",
    "points(observed_recalls,observed_FPs,col=col[names(observed_FPs)],pch=16,cex=2)\n",
    "\n",
    "s0fun<-splinefun(baselineRecalls,baselineNEratio)\n",
    "\n",
    "par(mar=c(8,6,4,4))\n",
    "barplot(observed_FPs/s0fun(observed_recalls),col=col[names(observed_FPs)],\n",
    "        las=2,border=NA,ylab='False CFs at fixed Recall of TPs / baseline')\n",
    "abline(h=1,lty=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the first plot, we show the threshold required by the baseline classifier to attain the observed recalls for each method. Although the 90-th percentile variants have lower threshold, they score higher when considering the median dependent cell lines of their CFs compared to the baseline classifier as shown in the second plot. This means that the predicted CF genes tend to be essential in a larger number of cell lines compared to CEN-tools and ADaM."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "baselineRecalls<-100*unlist(lapply(baselineCFs,function(x){\n",
    "  length(intersect(x,signatures))/length(signatures)\n",
    "}))\n",
    "\n",
    "par(mar=c(8,6,4,2))\n",
    "plot(baselineRecalls,xlab='Baseline predictor of CF genes (essentials in >= n cell lines)',\n",
    "     ylab='% Recall of prior known essential genes not used for training',pch=16)\n",
    "\n",
    "s0fun<-splinefun(baselineRecalls,1:length(baselineRecalls))\n",
    "abline(v=s0fun(observed_recalls),col=col[names(observed_recalls)],lwd=2)\n",
    "\n",
    "baseline_dep_cell_lines<-s0fun(observed_recalls)\n",
    "\n",
    "median_n_dependent_cell_lines<-\n",
    "  unlist(lapply(fCFs_sets,function(x){median(rowSums(bdep[intersect(x,geneSet),]))}))\n",
    "\n",
    "barplot(median_n_dependent_cell_lines/baseline_dep_cell_lines,col=col[names(median_n_dependent_cell_lines)],\n",
    "        las=2, border=NA,main = 'Median dependent cell lines on the CFs / baseline', ylab = 'ratio')\n",
    "abline(h=1,lty=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the first plot, we show the median fitness effect of the CF genes predicted by each method relative to the baseline predictor. CF genes predicted by the 90-th percentile variants have higher fitness score compared to ADaM and CEN-tools. This can be explained by the lower stringency of these methods that include genes with higher fitness score, therefore diluting the signal. <br>\n",
    "In the second plot, we show the number of CF genes for each method compared to the baseline predictor."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "observed_sizes<-lapply(CFs_sets,length)\n",
    "\n",
    "median_dep <- unlist(lapply(CFs_sets,\n",
    "                function(x){median(apply(scaled_depFC[intersect(x,geneSet_noBAG),],1,median))}))\n",
    "\n",
    "baseline_dep <- unlist(lapply(baselineCFs[round(s0fun(observed_recalls))],\n",
    "                function(x){median(apply(scaled_depFC[intersect(x,geneSet_noBAG),],1,median))}))\n",
    "\n",
    "par(mar=c(8,6,4,2))\n",
    "barplot(median_dep/baseline_dep,col=col[names(median_dep)],\n",
    "        las=2,main = 'Median CFs fitness effect / baseline', border=NA, ylab = 'ratio')\n",
    "abline(h=1,lty=2)\n",
    "\n",
    "baseline_size<-unlist(lapply(baselineCFs[round(s0fun(observed_recalls))],length))\n",
    "observed_sizes<-unlist(observed_sizes)\n",
    "\n",
    "barplot(observed_sizes/baseline_size,col=col[names(observed_sizes)],\n",
    "        las=2,main = 'n. CFs / baseline',border=NA, ylab = 'ratio')\n",
    "abline(h=1,lty=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison between methods accounting for BAGEL genes\n",
    "We perform the same pipeline shown above, but this time including the BAGEL gene sets used in the training phase of CEN-tools sets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Essentials genes obtained with CENtools using BAGEL curated genes as reference sets (default nº bin = 20)\n",
    "load('data/CENtools_Essentials.RData')\n",
    "\n",
    "## CENtools\n",
    "data(curated_BAGEL_essential)\n",
    "CENtools_SV1_B20Q2_CERES <- c(CENtools_ess, curated_BAGEL_essential)\n",
    "\n",
    "## CENtools Sharma et al, 2020\n",
    "load('data/BAGEL_v2_Essentials.RData')\n",
    "CENtools_2020 <- read.table(file = 'data/CenTools_essential_Sharma_et_al.txt',sep='\\t',stringsAsFactors = FALSE)$V1\n",
    "CENtools_2020 <- c(CENtools_2020, BAGEL_essential)\n",
    "\n",
    "## ADaM essential Behan et al., 2019\n",
    "load('data/ADaM_CFs_Behan_et_Al_2019.RData')\n",
    "ADaM_2019<-PanCancerCoreFitnessGenes\n",
    "\n",
    "col=c(\"#03B2C8\",\"#034DD9\",\"#E6AB02\",\"#1B9E77\",\"#337100\",\"#FC8D62\",'black',\n",
    "      \"#F4CAE4\",'purple',\"#CAB2D6\",\"#BEBADA\",\"#777892\")\n",
    "names(col) <- c(\"Hart_2015\",\"Hart_2017\",\"Behan_2019\",\"Sharma_2020\",\"CENtools\",\"ADaM\",\n",
    "              \"F_Perc_Cons\",\"F_Perc_avg\",\"F_Perc_Cons_lax\",\"F_Perc_slope\",\"F_Perc_AUC\",\"F_Perc_fixed\")\n",
    "\n",
    "##### Ability to identify novel hits\n",
    "CFs_sets<-list(Behan_2019 = ADaM_2019,\n",
    "               Sharma_2020 = CENtools_2020,\n",
    "               CENtools = CENtools_SV1_B20Q2_CERES,\n",
    "               ADaM = ADaM,\n",
    "               F_Perc_avg = Perc_avg,\n",
    "               F_Perc_Cons_lax = Perc_Consensus,\n",
    "               F_Perc_slope = Perc_slope,\n",
    "               F_Perc_AUC = Perc_AUC,\n",
    "               F_Perc_fixed = Perc_fixed)\n",
    "\n",
    "## Assembling a set of prior known essential genes that are not included in the\n",
    "## training sets used by CENtools and/or ADaM\n",
    "signatures<-list(DNA_REPLICATION=EssGenes.DNA_REPLICATION_cons,\n",
    "                 RNA_POLYMERASE=EssGenes.KEGG_rna_polymerase,\n",
    "                 PROTEASOME=EssGenes.PROTEASOME_cons,\n",
    "                 SPLICEOSOME=EssGenes.SPLICEOSOME_cons,\n",
    "                 RIBOSOMAL_PROTS=EssGenes.ribosomalProteins,\n",
    "                 HISTONES=EssGenes.HISTONES)\n",
    "\n",
    "signatures<-unique(unlist(signatures))\n",
    "\n",
    "## Adding additional signatures in Pacini et al, 2020\n",
    "allRefEss<-unique(c(Kegg.DNArep,Kegg.Proteasome,Kegg.Ribosome,Kegg.RNApoly,Kegg.Spliceosome,Histones))\n",
    "\n",
    "signatures<-union(signatures,allRefEss)\n",
    "signatures <- intersect(signatures,geneSet)\n",
    "\n",
    "## baseline predictor\n",
    "baselineCFs <- lapply(1:ncol(bdep),function(x){\n",
    "  names(which(rowSums(bdep)>=x))\n",
    "})\n",
    "\n",
    "baselineSizes<-unlist(lapply(baselineCFs,length))\n",
    "\n",
    "baselineRecalls<-100*unlist(lapply(baselineCFs,function(x){\n",
    "  length(intersect(x,signatures))/length(signatures)\n",
    "}))\n",
    "\n",
    "## genes not expressed\n",
    "lowlyExp<-names(which(rowSums(CMP_RNAseq<0.01,na.rm=TRUE)>=ncol(CMP_RNAseq)))\n",
    "\n",
    "## dependency associated with biomarkers\n",
    "FPs <- sort(unique(unlist(read.table('data/dependency_with_biomarkers.txt',stringsAsFactors = FALSE))))\n",
    "FPs <- union(FPs,lowlyExp)\n",
    "FPs <- intersect(FPs,geneSet_noBAG)\n",
    "\n",
    "baselineNEratio<-100*unlist(lapply(baselineCFs,function(x){\n",
    "  length(intersect(x,FPs))/length(FPs)\n",
    "}))\n",
    "\n",
    "plot(baselineSizes,xlab='Baseline predictor of CF genes (essentials in >= n cell lines)', \n",
    "     ylab='n. of predicted CF genes',pch=16,log='y', main = \"BAGEL genes included\")\n",
    "\n",
    "plot(baselineRecalls,xlab='Baseline predictor of CF genes (essentials in >= n cell lines)',\n",
    "     ylab='% Recall of prior known essential genes not used for training',pch=16)\n",
    "\n",
    "plot(baselineNEratio,xlab='Baseline predictor of CF genes (essentials in >= n cell lines)',\n",
    "     ylab='% Recall of selective essentials (associated with a biomarker)',pch=16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fCFs_sets <- CFs_sets\n",
    "\n",
    "observed_recalls<-100*unlist(lapply(fCFs_sets,\n",
    "                                    function(x){length(intersect(x,signatures))/length(signatures)}))\n",
    "\n",
    "observed_FPs<-100*unlist(lapply(fCFs_sets,\n",
    "                                function(x){length(intersect(x,FPs))/length(FPs)}))\n",
    "\n",
    "par(mar=c(8,6,4,4))\n",
    "barplot(observed_recalls/max(baselineRecalls),col=col[names(observed_recalls)],\n",
    "        las=2,ylab='Recall / (max Recall)',border=NA,main = 'Prior known essential with training genes included')\n",
    "\n",
    "barplot(observed_FPs/max(baselineNEratio),col=col[names(observed_FPs)],\n",
    "        las=2,ylab='Recall / (max Recall)', border=NA, main = 'False positive CFs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot(spline(baselineRecalls,baselineNEratio),pch=16,\n",
    "     xlab='% Recall of prior known essentials',\n",
    "     ylab='% Recall of selective essentials (associated with a biomarker)',type='l',lwd=4,\n",
    "     xlim=c(40,70),ylim=c(0,6))\n",
    "\n",
    "points(observed_recalls,observed_FPs,col=col[names(observed_FPs)],pch=16,cex=2)\n",
    "\n",
    "s0fun<-splinefun(baselineRecalls,baselineNEratio)\n",
    "\n",
    "par(mar=c(8,6,4,4))\n",
    "barplot(observed_FPs/s0fun(observed_recalls),col=col[names(observed_FPs)],\n",
    "        las=2,border=NA,ylab='False CFs at fixed Recall of TPs / baseline')\n",
    "abline(h=1,lty=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "baselineRecalls<-100*unlist(lapply(baselineCFs,function(x){\n",
    "  length(intersect(x,signatures))/length(signatures)\n",
    "}))\n",
    "\n",
    "par(mar=c(8,6,4,2))\n",
    "plot(baselineRecalls,xlab='Baseline predictor of CF genes (essentials in >= n cell lines)',\n",
    "     ylab='% Recall of prior known essential genes',pch=16)\n",
    "\n",
    "s0fun<-splinefun(baselineRecalls,1:length(baselineRecalls))\n",
    "abline(v=s0fun(observed_recalls),col=col[names(observed_recalls)],lwd=2)\n",
    "\n",
    "baseline_dep_cell_lines<-s0fun(observed_recalls)\n",
    "\n",
    "median_n_dependent_cell_lines<-\n",
    "  unlist(lapply(fCFs_sets,function(x){median(rowSums(bdep[intersect(x,geneSet),]))}))\n",
    "\n",
    "barplot(median_n_dependent_cell_lines/baseline_dep_cell_lines,col=col[names(median_n_dependent_cell_lines)],\n",
    "        las=2, border=NA,main = 'Median dependent cell lines on the CFs / baseline', ylab = 'ratio')\n",
    "abline(h=1,lty=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "observed_sizes<-lapply(CFs_sets,length)\n",
    "\n",
    "median_dep <- unlist(lapply(CFs_sets,\n",
    "                function(x){median(apply(scaled_depFC[intersect(x,geneSet),],1,median))}))\n",
    "\n",
    "baseline_dep <- unlist(lapply(baselineCFs[round(s0fun(observed_recalls))],\n",
    "                function(x){median(apply(scaled_depFC[intersect(x,geneSet),],1,median))}))\n",
    "\n",
    "par(mar=c(8,6,4,2))\n",
    "barplot(median_dep/baseline_dep,col=col[names(median_dep)],\n",
    "        las=2,main = 'Median CFs fitness effect / baseline', border=NA, ylab = 'ratio')\n",
    "abline(h=1,lty=2)\n",
    "\n",
    "baseline_size<-unlist(lapply(baselineCFs[round(s0fun(observed_recalls))],length))\n",
    "observed_sizes<-unlist(observed_sizes)\n",
    "\n",
    "barplot(observed_sizes/baseline_size,col=col[names(observed_sizes)],\n",
    "        las=2,main = 'n. CFs / baseline',border=NA, ylab = 'ratio')\n",
    "abline(h=1,lty=2)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
